pkg/build/api/validation/validation.go

```diff
@@ -253,9 +255,32 @@ func validateImageSource(imageSource *buildapi.ImageSource) fielderrors.Validati
 	if len(imageSource.Paths) == 0 {
 		allErrs = append(allErrs, fielderrors.NewFieldRequired("paths"))
 	}
+	for i, path := range imageSource.Paths {
+		allErrs = append(allErrs, validateImageSourcePath(path).PrefixIndex(i).Prefix("paths")...)
+
+	}
 	return allErrs
 }
 
+func validateImageSourcePath(imagePath buildapi.ImageSourcePath) fielderrors.ValidationErrorList {
+	allErrs := fielderrors.ValidationErrorList{}
+	if len(imagePath.SourcePath) == 0 {
+		allErrs = append(allErrs, fielderrors.NewFieldRequired("sourcePath"))
+	}
+	if len(imagePath.DestinationDir) == 0 {
+		allErrs = append(allErrs, fielderrors.NewFieldRequired("destinationDir"))
+	}
+	if !filepath.IsAbs(imagePath.SourcePath) {
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("sourcePath", imagePath.SourcePath, "must be an absolute path"))
+	}
+	if filepath.IsAbs(imagePath.DestinationDir) {
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("destinationDir", imagePath.DestinationDir, "must be a relative path"))
+	}
+	if strings.HasPrefix(path.Clean(imagePath.DestinationDir), "..") {
```

Comment:
It's not system dependent: https://golang.org/pkg/path/.

It has no effect in backslash separated strings (e.g. Windows paths): http://play.golang.org/p/Retd8NjjXj
IOW, the cleaned version of `"foo\..\..\..\....."` is `"foo\..\..\..\....."` itself.